File: backend/workflows/serializers.py
"""
Django REST Framework serializers for workflows
"""
from rest_framework import serializers
from .models import Workflow, WorkflowExecution, Credential, ExportedWorkflow


class WorkflowSerializer(serializers.ModelSerializer):
    """Serializer for Workflow model"""
    
    class Meta:
        model = Workflow
        fields = ['id', 'name', 'description', 'nodes', 'edges', 'created_at', 'updated_at', 'is_active']
        read_only_fields = ['id', 'created_at', 'updated_at']


class WorkflowExecutionSerializer(serializers.ModelSerializer):
    """Serializer for WorkflowExecution model"""
    
    class Meta:
        model = WorkflowExecution
        fields = [
            'id', 'workflow', 'status', 'started_at', 'finished_at',
            'execution_order', 'node_states', 'errors', 'trigger_data'
        ]
        read_only_fields = ['id', 'started_at']


class CredentialSerializer(serializers.ModelSerializer):
    """Serializer for Credential model"""
    
    class Meta:
        model = Credential
        fields = ['id', 'name', 'credential_type', 'data', 'created_at', 'updated_at']
        read_only_fields = ['id', 'created_at', 'updated_at']


class ExecuteWorkflowSerializer(serializers.Serializer):
    """Serializer for workflow execution request"""
    trigger_data = serializers.JSONField(required=False, default=dict)
    start_node_id = serializers.CharField(required=False, allow_null=True)
    credentials = serializers.JSONField(required=False, default=dict)


class ExecuteNodeSerializer(serializers.Serializer):
    """Serializer for single node execution request"""
    node_id = serializers.CharField(required=True)
    trigger_data = serializers.JSONField(required=False, default=dict)
    credentials = serializers.JSONField(required=False, default=dict)


class ExportedWorkflowSerializer(serializers.ModelSerializer):
    """Serializer for ExportedWorkflow model"""
    
    class Meta:
        model = ExportedWorkflow
        fields = [
            'id', 'name', 'description', 'version', 'export_type',
            'nodes', 'edges', 'tags', 'category', 'author',
            'is_public', 'is_featured', 'created_at', 'updated_at',
            'exported_at', 'download_count', 'import_count'
        ]
        read_only_fields = ['id', 'created_at', 'updated_at', 'exported_at', 'download_count', 'import_count']


class ExportedWorkflowCreateSerializer(serializers.ModelSerializer):
    """Serializer for creating exported workflows"""
    
    class Meta:
        model = ExportedWorkflow
        fields = [
            'name', 'description', 'version', 'export_type',
            'nodes', 'edges', 'tags', 'category', 'author',
            'is_public', 'is_featured'
        ]
    
    def validate_nodes(self, value):
        """Validate nodes data"""
        if not isinstance(value, list):
            raise serializers.ValidationError("Nodes must be a list")
        return value
    
    def validate_edges(self, value):
        """Validate edges data"""
        if not isinstance(value, list):
            raise serializers.ValidationError("Edges must be a list")
        return value


class ExportedWorkflowListSerializer(serializers.ModelSerializer):
    """Lightweight serializer for exported workflow lists"""
    
    class Meta:
        model = ExportedWorkflow
        fields = [
            'id', 'name', 'description', 'version', 'export_type',
            'category', 'author', 'is_public', 'is_featured',
            'exported_at', 'download_count', 'import_count'
        ]
        read_only_fields = ['id', 'exported_at', 'download_count', 'import_count']

File: web/src/lib/serializers.ts
/**
 * Serialization helpers that convert Prisma records into the snake_case
 * shapes the existing React frontend expects (ported from DRF serializers).
 */
import type {
  User,
  Workflow,
  WorkflowExecution,
  Credential,
  ExportedWorkflow,
  UIBuilderProject,
} from "@prisma/client";

export function serializeUser(user: User) {
  return {
    id: user.id,
    username: user.username,
    email: user.email ?? "",
    first_name: user.firstName ?? "",
    last_name: user.lastName ?? "",
    date_joined: user.createdAt,
  };
}

export function serializeWorkflow(w: Workflow) {
  return {
    id: w.id,
    name: w.name,
    description: w.description,
    nodes: w.nodes,
    edges: w.edges,
    created_at: w.createdAt,
    updated_at: w.updatedAt,
    is_active: w.isActive,
  };
}

export function serializeExecution(e: WorkflowExecution) {
  return {
    id: e.id,
    workflow: e.workflowId,
    status: e.status,
    started_at: e.startedAt,
    finished_at: e.finishedAt,
    execution_order: e.executionOrder,
    node_states: e.nodeStates,
    errors: e.errors,
    trigger_data: e.triggerData,
  };
}

export function serializeCredential(c: Credential) {
  return {
    id: c.id,
    name: c.name,
    credential_type: c.credentialType,
    data: c.data,
    created_at: c.createdAt,
    updated_at: c.updatedAt,
  };
}

export function serializeExportedWorkflow(x: ExportedWorkflow) {
  return {
    id: x.id,
    user: x.userId,
    name: x.name,
    description: x.description,
    version: x.version,
    export_type: x.exportType,
    nodes: x.nodes,
    edges: x.edges,
    tags: x.tags,
    category: x.category,
    author: x.author,
    is_public: x.isPublic,
    is_featured: x.isFeatured,
    created_at: x.createdAt,
    updated_at: x.updatedAt,
    exported_at: x.exportedAt,
    download_count: x.downloadCount,
    import_count: x.importCount,
  };
}

export function serializeUIProject(p: UIBuilderProject) {
  return {
    id: p.id,
    project_name: p.projectName,
    description: p.description,
    components: p.components,
    styles: p.styles,
    assets: p.assets,
    created_at: p.createdAt,
    updated_at: p.updatedAt,
    is_active: p.isActive,
  };
}

/** DRF-style paginated list response */
export function paginated<T>(results: T[]) {
  return {
    count: results.length,
    next: null,
    previous: null,
    results,
  };
}
